docs/design_proposals/artifact-dependencies.md

```diff
@@ -0,0 +1,356 @@
+}
+
+func (a *artifactChanModel) markSuccess() {
+	// closing channel notifies all listeners waiting for this build that it succeeded
+	close(a.status.success)
+}
+
+func (a *artifactChanModel) markFailure() {
+	// closing channel notifies all listeners waiting for this build that it failed
+	close(a.status.failure)
+}
+func (a *artifactChanModel) waitForDependencies(ctx context.Context) error {
+	for _, depStatus := range a.requiredArtifactChans {
+		// wait for required builds to complete
+		select {
+		case <-ctx.Done():
+		case <-depStatus.failure:
+			return fmt.Errorf("failed to build required artifact: %q", depStatus.imageName)
+		case <-dep:
+		}
+	}
+    return nil
+}
+```
+
+Each artifact has a success and a failure channel that it closes once it completes building by calling either `markSuccess` or `markFailure` respectively. This notifies *all* listeners waiting for this artifact of a successful or failed build.
+
+Additionally it has a reference to the channels for each of its dependencies.
+Calling `waitForDependencies` ensures that all required artifacts' channels have already been closed and as such have finished building before the current artifact build starts.
+
+> *<ins>Alternative approach</ins>:  Another way to do this is to run any popular topologically sorting algorithm on the `Artifact` slice, treating it as a directed graph. However, we can get a simpler implementation at the expense of a few additional `goroutines` the way described above.*
+
+This class also provides an implementation of the interface `buildStatusRecorder` that should be safe for concurrent access.
+
+```go
+type buildStatusRecorder interface {
+  Record(imageName string, imageTag string, err error)
+  GetImageTag(imageName string) string
+}
+```
+
+Finally we have the only exported function in `scheduler.go` that orchestrates all the builds:
+
+```go
+func InOrder(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, existing []Artifact,  buildArtifact ArtifactBuilder, concurrency int) ([]Artifact, error) {
+```
+
+This function maintains an instance of `buildStatusRecorder` implementation and can pass it as an `ArtifactResolver` to the various `ArtifactBuilder`s while recording the status after each build completion.
+
+## Build concurrency
+
+Skaffold currently allows specifying the `concurrency` property in `build` which affects how many builds can be running at the same time. However it doesn't address the issue of certain builders (`jib` and `bazel`) not being safe for multiple concurrent runs against the same workspace or context. We can fix this also since we are reworking the build controller anyways. 
+
+We define a concept of lease on workspaces by preprocessing the list of artifacts. Each builder tries to acquire a lease on the context/workspace prior to starting the build. Only workspaces associated with concurrency-safe builders allot multiple leases, otherwise it assigns one lease at a time.
```

Comment:
What happens if we have several artifacts of different builder-type but same context?

Reply:
> `Only workspaces associated with concurrency-safe builders allot multiple leases`

So even if one builder is concurrency unsafe, it'll be single leases for that context.

Reply:
Sounds like we need a builder interface (rather than just the `ArtifactBuilder` function signature).  Can you sketch out what this interface would look like?

Reply:
Added